% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  sn-mathphys-num,
  lineno,
  twocolumn]{sn-jnl}



\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else  
    % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
% Make \paragraph and \subparagraph free-standing
\makeatletter
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}{
    \@ifstar
      \xxxParagraphStar
      \xxxParagraphNoStar
  }
  \newcommand{\xxxParagraphStar}[1]{\oldparagraph*{#1}\mbox{}}
  \newcommand{\xxxParagraphNoStar}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}{
    \@ifstar
      \xxxSubParagraphStar
      \xxxSubParagraphNoStar
  }
  \newcommand{\xxxSubParagraphStar}[1]{\oldsubparagraph*{#1}\mbox{}}
  \newcommand{\xxxSubParagraphNoStar}[1]{\oldsubparagraph{#1}\mbox{}}
\fi
\makeatother

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{241,243,245}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.40,0.45,0.13}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\BuiltInTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\ExtensionTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.28,0.35,0.67}{#1}}
\newcommand{\ImportTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.07,0.07,0.07}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}

\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother

%%%% Standard Packages

\usepackage{graphicx}%
\usepackage{multirow}%
\usepackage{amsmath,amssymb,amsfonts}%
\usepackage{amsthm}%
\usepackage{mathrsfs}%
\usepackage[title]{appendix}%
\usepackage[dvipsnames]{xcolor}%
\usepackage{textcomp}%
\usepackage{manyfoot}%
\usepackage{booktabs}%
\usepackage{algorithm}%
\usepackage{algorithmicx}%
\usepackage{algpseudocode}%
\usepackage{listings}%

%%%%

\raggedbottom
\usepackage{nameref}
\usepackage{placeins}
\usepackage[labelfont=bf]{caption}
% \usepackage{showframe}
% \usepackage{layout}
\allowdisplaybreaks
\makeatletter
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{plain}
\@ifundefined{c@chapter}{\newfloat{suppfig}{h}{loextfig}}{\newfloat{suppfig}{h}{loextfig}[chapter]}
\floatname{suppfig}{Extended Data Fig.}
\newcommand*\listofsuppfigs{\listof{suppfig}{List of Extended Data Figures}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother

\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}

\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={Pyro -Velocity: Probabilistic RNA Velocity inference from single-cell data},
  pdfkeywords={RNA velocity, single-cell RNA sequencing, probabilistic
generative modeling, cell fate, developmental trajectory inference},
  colorlinks=true,
  linkcolor={NavyBlue},
  filecolor={Maroon},
  citecolor={NavyBlue},
  urlcolor={NavyBlue},
  pdfcreator={LaTeX via pandoc}}


\title[Pyro -Velocity: Probabilistic RNA Velocity inference from
single-cell data]{Pyro -Velocity: Probabilistic RNA Velocity inference
from single-cell data}

% author setup
\author[1,2,3,4,5]{\fnm{Qian} \sur{Qin}}\author[5,7]{\fnm{Eli} \sur{Bingham}}\author*[6]{\fnm{Gioele La} \sur{Manno}}\email{gioele.lamanno@epfl.ch}\author*[1,2,3,4]{\fnm{David M.} \sur{Langenau}}\email{dlangenau@mgh.harvard.edu}\author*[1,2,5]{\fnm{Luca} \sur{Pinello}}\email{lpinello@mgh.harvard.edu}
% affil setup
\affil[1]{\orgdiv{Molecular Pathology Unit}, \orgname{Massachusetts
General Hospital Research Institute, Charlestown, MA}}
\affil[2]{\orgdiv{Massachusetts General Hospital Cancer
Center}, \orgname{Harvard Medical School, Charlestown, MA}}
\affil[3]{\orgdiv{Center for Regenerative
Medicine}, \orgname{Massachusetts General Hospital, Boston, MA}}
\affil[4]{\orgdiv{Harvard Stem Cell Institute}, \orgname{Harvard
University, Cambridge, MA}}
\affil[5]{\orgdiv{Broad Institute of MIT and
Harvard}, \orgname{Cambridge, MA}}
\affil[6]{\orgdiv{Brain and Mind Institute, School of Life Sciences,
École Polytechnique Fédérale de Lausanne (EPFL)}, \orgname{Laboratory of
Neurodevelopmental Systems Biology, Lausanne, Switzerland}}
\affil[7]{\orgdiv{Basis Research Institute}, \orgname{Cambridge, MA}}

% abstract 

\abstract{Single-cell RNA velocity has dramatically advanced our ability
to model cellular differentiation and cell fate decisions. However,
current preprocessing choices and model assumptions often lead to errors
in assigning developmental trajectories. Here, we develop,
Pyro -Velocity, a Bayesian, generative, and multivariate RNA Velocity
model to estimate the uncertainty of cell future states. This approach
models raw sequencing counts with the synchronized cell time across all
expressed genes to provide quantifiable and improved information on cell
fate choices and developmental trajectory dynamics.}

% keywords
\keywords{RNA velocity,  single-cell RNA sequencing,  probabilistic
generative modeling,  cell fate,  developmental trajectory inference}

\begin{document}
\maketitle
% \clearpage
% \layout
\newpage{}

\renewcommand*\contentsname{Contents}
{
\hypersetup{linkcolor=NavyBlue}
\setcounter{tocdepth}{2}
\tableofcontents
}

\newpage{}

\section*{Main text}\label{sec-main}
\addcontentsline{toc}{section}{Main text}

RNA velocity is a powerful computational framework to estimate the time
derivative of gene expression
\citep{La_Manno2018-lj, Svensson2018-vk, Qiu2022-dj, Bergen2020-pj}. The
framework has been used to study developmental cell lineage trajectories
over a broad spectrum of biological processes, such as haematopoiesis
\citep{Qiu2022-dj}, pancreas development \citep{Bastidas-Ponce2019-lf},
and neural progenitors differentiation \citep{Bergen2020-pj}. Yet,
recent state-of-the-art RNA Velocity approaches such as velocyto and
scVelo \citep{La_Manno2018-lj} have several limitations. First, they may
predict developmental trajectories that do not exist in nature, as it
has been shown in the analysis of fully mature blood cells
\citep{Zheng2017-bz, Bergen2021-qz}. Second, data preprocessing steps
(e.g., selection of top variable genes, dimensionality reduction) affect
velocity estimates that are difficult to characterize and might lead to
artifactual predictions. In addition, non-linear spliced and unspliced
imputation procedures could distort phase portrait geometry and
recovered trajectories (Fig.~\ref{fig-model}a). Third, most RNA velocity
algorithms perform estimates independently for each gene, and therefore
cannot exploit the natural interdependence of gene expression. Fourth, a
unified description of the temporal relation between cells can only be
assigned based on postprocessing steps that aggregate heuristically the
gene-wise fits into a shared latent time (Fig.~\ref{fig-model}a).
Finally, current RNA velocity frameworks only output point estimates of
the velocity vector field and do not provide uncertainty estimation.
Importantly, the absence of statistical confidence estimates increases
false discoveries as it leaves users without a way to evaluate the
robustness of the methods' outputs. To overcome these limitations, we
developed a probabilistic generative model, named \emph{Pyro -Velocity},
that estimates RNA dynamics and infers cell trajectories using Pyro
\citep{Bingham2018-id}, a robust and efficient probabilistic programming
language.

Pyro -Velocity recasts the velocity estimation problem into a latent
variable posterior probability inference
(\hyperref[sec-methods]{Methods}). The proposed model is generative and
fully Bayesian, with the different parameters considered as latent
random variables. Each parameter is assigned a prior distribution, and
the model is then conditioned on the raw data (both spliced and
unspliced molecule raw counts) to estimate posterior probability
distributions and to enable uncertainty estimation of velocity. Central
to the Pyro -Velocity model is a \emph{shared latent time} \((t)\): a
global latent variable placing each cell at a particular instant of the
modeled dynamical process. From this latent time, the expectation of
spliced and unspliced RNA levels for each gene can be computed knowing
gene-specific kinetics parameters (transcriptional \((\alpha)\),
splicing \((\beta)\), and degradation rates (\(\gamma\)) and two
gene-specific switching times \((t_{0}^{k})\), and gene-specific
transcriptional state \(k\)) and integrating the RNA velocity
differential equations analogously to what proposed in the dynamical RNA
velocity model by Bergen et al. \citep{Bergen2020-pj, Li2021-qa}.
However, we innovate the structure and formulation of the latent time
given that our model is set to jointly fit the entire raw data matrix,
not individual genes. We approach the task of estimating the posterior
distribution by restricting the search within a family of a variational
distribution approximating the true posterior, and taking advantage of
the automatic differentiation variational inference (ADVI) capabilities
of Pyro \citep{Bingham2018-id, Kucukelbir2016-bk}
(Fig.~\ref{fig-model}a, \hyperref[sec-methods]{Methods}). Notably, our
generative Bayesian model enables to sample from the posterior
distributions of the latent and dependent variables, allowing
uncertainty estimation of phase portraits, high dimensional velocity
vector field, and shared latent time. Around this model, we built
several visualizations to assess uncertainty and functions to recover
putative cell fate priming genes and to investigate how the denoised
gene expression recapitulates known cell-type annotations and their
developmental order (Fig.~\ref{fig-model}b,
\hyperref[sec-methods]{Methods}). Although several RNA velocity methods
\citep{Gayoso2024-fz, Gorin2022-il, Gu2022-rh, Cui2022-em, Gao2022-jq}
have been recently proposed, here we showcase the advantages and
innovations of Pyro -Velocity by comparing it with the current
state-of-the-art dynamical RNA velocity model proposed by Bergen et al.
\citep{Bergen2020-pj}, and implemented in scVelo.

One major biological hurdle for current formulations of RNA Velocity
models has been miscalling cell trajectories in fully differentiated
cell populations that lack dynamic cell state transitions. One typical
example being human peripheral blood mononuclear cells (PBMCs)
\citep{Zheng2017-bz}. In this dataset, 65,877 cells were analyzed using
the 10x platform that comprised 11 fully differentiated, mature immune
cell types. This dataset lacked stem and progenitor cells or other
signatures of an undergoing dynamical differentiation; thus, no
consistent velocity flow should be detected. We reanalyzed this dataset
using scVelo and Pyro -Velocity using a common set of genes for a fair
comparison. Using scVelo, we reproduced the original results showing
incorrect cell trajectory assignments (Fig.~\ref{fig-model-results}a)
\citep{Bergen2021-qz}. ScVelo uncovered an errant, strong linear
trajectory across a subset of mature cell types \citep{Zheng2017-bz}
that is exasperated by the inability to confidently assess lineage
directionality or a baseline to assess significance. On the contrary,
Pyro -Velocity failed to detect high-confidence trajectories in the
mature blood cell states, consistent with what is known about the
biology underlying these cells (Fig.~\ref{fig-model-results}a, panel 2
vs.~panel 3). Importantly, the Bayesian nature of Pyro -Velocity enabled
the generation of uncertainty estimates that can be easily visualized as
likely vectors for individual cells or averaged across cells on a
constantly spaced grid on the cell embedding (\emph{averaged vector
field}).

Pyro -Velocity was also able to define well-known developmental cell
hierarchies in a scRNA-seq dataset of E15.5 mouse pancreas
(Fig.~\ref{fig-model-results}b-e, \(n=3,696\) cells comprising \(8\)
cell types (Fig.~\ref{fig-model-results}b, panel 1)), identifying cell
trajectories originating from ductal progenitor cells and culminated in
the production of mature Alpha, Beta, Delta, and Epsilon cells
\citep{Bastidas-Ponce2019-lf}. Though the scVelo and Pyro -Velocity
models have qualitatively comparable vector fields, the results
generated by scVelo were greatly influenced by gene selection,
neighboring cell number for gene expression smoothing, and cannot
provide confidence assessment of lineage directionality
(Extended Data Fig.~\ref{extfig-model-results}a). In addition,
Pyro -Velocity uncertainty visualizations uncovered that differentiated
cells had a larger vector field uncertainty than stem and progenitor
cells, as would be expected for terminally differentiated cell states
where progression and lineage trajectories would not be anticipated.
Pyro -Velocity also recovered a developmental order in close agreement
with what was estimated by Cytotrace (Spearman's
\(r=0.95, \; \text{p-value} <
10^{-10}\) between latent time and developmental
order)\citep{Gulati2020-xq} - a state-of-the-art orthogonal method used
to predict cell differentiation based on the number of expressed genes
per cell.

We next used Pyro -Velocity to unbiasedly uncover potential cell fate
determinant markers genes within the E15.5 mouse pancreas. To this end,
we first selected the top 300 genes with the best goodness of fit
(negative mean absolute errors across cells). Next, we ranked them based
on the positive (or negative) correlation between inferred shared time
and corresponding denoised spliced expression per cell
(Fig.~\ref{fig-model-results}d, \hyperref[sec-methods]{Methods}). This
procedure successfully recovered key developmental genes that are
activated (or repressed) during endocrinogenesis
(Fig.~\ref{fig-model-results}d-e,
Extended Data Fig.~\ref{extfig-model-results}). Among the top positively
correlated genes identified by Pyro -Velocity, endocrine Beta cells
expressed insulin 1 and 2 (Ins1, Ins2) and Nnat, which is a key
regulator Beta cell insulin generation \citep{Millership2018-vd} and
Ppp1r1a which is a marker of Beta cells \citep{Jiang2013-pt}. The
ranking procedure proposed by scVelo failed to identify Ins1, Ins2,
Nnat, and Ppp1r1a as lineage-restricted genes in their top 50 ranked
gene list. The inability of scVelo to identify these key developmental
genes could be likely explained by the kNN smoothing of read count and
filtering steps (Supplementary Table 1, 2). Visual inspection of these
four top genes by Pyro -Velocity revealed a single activation phase
during cell type transitions based on phase portrait curves. We also
propose their visualization through ``rainbow plots'\,' to illustrate
denoised single gene spliced expression across shared time and to
investigate the inferred cell order based on known annotations. Further,
UMAP embedding of cells colored based on denoised gene expression
confirmed the cell fate markers' smooth and progressive activation (or
repression) towards endocrine cells (Fig.~\ref{fig-model-results}e,
Extended Data Fig.~\ref{extfig-model-results}).

Ground truth velocity is not a quantity we can easily assess. Therefore,
no robust and quantitative benchmarking of RNA velocity methods has been
proposed. However, recent single-cell assays allow recording lineage
relationships and can be used to assess the ability of different methods
in recapitulating cell fate progression. Therefore, we next
systematically compared scVelo and Pyro -Velocity ability to recover
single cell fates in Lineage And RNA RecoverY (LARRY) barcoded
hematopoietic cells. LARRY leverages unique lentiviral barcodes inserted
into 3' untranslated region (UTR) of GFP reporter that is integrated as
a single copy into parental cells. Then progeny can be clonally traced
to follow the descendent cell fates over time. LARRY has been recently
applied to in vitro differentiation of human blood and accurately
recovered expected cell fate and lineage trajectories
\citep{Weinreb2020-mv}. This dataset sampled differentiation over the
course of 2, 4, and 6 days and contained \(49,302\) cells that each
could be traced with at least one barcode. Because LARRY couples
single-cell transcriptomics and direct cell lineage tracing, this data
set provides a unique opportunity to benchmark RNA Velocity
quantitatively and to provide insights into the correct interpretation
of recovered velocity vector fields and latent cell times. To
quantitatively assess the agreement between observed and predicted
fates, we considered two metrics: 1. average cosine similarity (cs)
between vectors from the RNA Velocity vector field and the clonal
progression vector field, i.e., a vector field derived directly from the
lineage data (\hyperref[sec-methods]{Methods}), 2. Spearman's
correlation \((\rho)\) between cell latent (shared) time and fate
potency scores derived from Cytotrace and Cospar. Cospar, our gold
standard, is a state-of-the-art method designed explicitly for
predicting fate potency based on LARRY data, while Cytotrace leverages
only transcriptomic data. We selected cell state changes associated with
monocyte and neutrophil lineages because both lineages have the largest
numbers of cells in the scRNA sequencing data set and well-known cell
fate trajectories \citep{Petukhov2018-nr}. As expected, the
Pyro -Velocity vector field accurately predicted cell state trajectories
from hematopoietic progenitor cells into mature monocytes or neutrophils
and had high cosine similarity between the inferred vector field and
clonal progression vector field assigned by LARRY (monocytes
\(cs=0.49\), neutrophils \(cs=0.38\)). High Spearman's correlation was
also seen between shared latent time and Cospar fate potency (monocytes
\(\rho=0.60\), neutrophils \(\rho=0.74\)) or shared latent time with
Cytotrace fate potency (monocytes \(\rho=0.925\), neutrophils
\(\rho=0.925\)). By contrast, scVelo generated vector fields with
reversed order or low correlation (cosine similarity: \(-0.38\),
monocytes, cosine similarity: \(0.32\), neutrophils). Moreover, Cospar
fate potency scores were negative (\(\rho=-0.59\), monocytes) or
non-significantly correlated (\(\rho=-0.07\), neutrophils). A similar
low correlation was observed for scVelo and Cytotrace fate potency
scores (monocytes \(\rho=0.309\), neutrophils \(\rho=0.751\))
(Fig.~\ref{fig-lineage-tracing}a-b,
Extended Data Fig.~\ref{extfig-lineage-tracing}c).

Next, we evaluated the ability of the Pyro -Velocity models to recover
bi-fate or multi-fate trajectories by considering clones belonging to
monocyte and neutrophil lineages or using the entire dataset. Not
surprisingly, both methods showed their limitations, given that they are
not designed to model and capture multiple fate trajectories
(Extended Data Fig.~\ref{extfig-lineage-tracing-2}c-d). The recovered
vector fields and cell latent (shared) time show inconsistent directions
for both methods. However, Pyro -Velocity identified higher uncertainty
for some of the problematic regions. Furthermore, we discovered that
vector field and shared time estimation could be significantly improved
by extending the original kinetic model by allowing each gene a time lag
for the transcriptional activation and a baseline expression level
potentially different than 0 (Pyro -Velocity Model 2,
\hyperref[sec-methods]{Methods}) (Fig.~\ref{fig-lineage-tracing}c,d,
Extended Data Fig.~\ref{extfig-graphical-models}b,c). This extended
model recapitulated a vector field consistent with the clonal
progression vector field for most branches with some spurious
predictions within the intermediate cell state between stem cells and
Megakaryocytes and most terminal Neutrophils (bi-fate cosine similarity:
\(0.62\), multi-fate cosine similarity: \(0.63\)). However, the proposed
uncertainty estimation accurately flagged these subpopulations to warn
users of potential false discoveries. Notably, this extended model
significantly improved the correlation of cell shared time with the
clonal fate potency (bi-fate \(\rho=0.65\), multi-fate \(\rho=0.74\))
and also improved the correlation with Cytotrace fate potency (bi-fate
\(\rho=0.931\), multi-fate \(\rho=0.614\)). Remarkably, the shared time
per cell from Pyro -Velocity also outperformed Cytotrace for
developmental order prediction
(Extended Data Fig.~\ref{extfig-lineage-tracing}a,c). Importantly, this
extended model retained its ability to accurately identify cell lineage
trajectories and shared time for all the other datasets analyzed above
(Extended Data Fig.~\ref{extfig-model-2-result}). Taken together, our
approach using LARRY barcoded datasets established the first
quantitative benchmark of RNA velocity-based cell fate trajectory
inference and provides an exciting new framework for continued
optimization of multi-fate predictions in the future.

In conclusion, Pyro -Velocity provides the first probabilistic,
interpretable, end-to-end Bayesian inference framework for RNA Velocity
that learns cell dynamics and recovers cell fate from multimodal raw
read counts. This approach provides uncertainty estimations of cell fate
trajectories and novel visualization approaches to mitigate false
discoveries. Pyro -Velocity recovered shared time and vector fields that
are recapitulated in true developmental order and outperformed
state-of-the-art methods like Cytotrace and scVelo
(Extended Data Fig.~\ref{extfig-lineage-tracing}a).

\section*{References}\label{references}
\addcontentsline{toc}{section}{References}

\renewcommand{\bibsection}{}
\bibliography{bibliography.bib}

\FloatBarrier

\newpage{}

\onecolumn

\section{Figures}\label{figures}

\begin{figure}

\centering{

\includegraphics{figures/v2/Figure1.pdf}

}

\caption{\label{fig-model}Pyro -Velocity is a fully generative Bayesian
method with uncertainty estimation of velocity vector fields and shared
latent time, based on raw counts and without ad-hoc preprocessing steps.
\textbf{a}, Schematic comparing current RNA Velocity and Pyro -Velocity
workflows. Both methods take in input spliced and unspliced read count
tables and output RNA velocity vector fields and latent (shared) time.
Previous methods require ad-hoc preprocessing steps, including the
arbitrary selection of top variable genes, number of principal
components, and kNN-pooling to impute both spliced and unspliced
expression within the PCA-reduced spliced expression space. In addition,
manually tuned ad-hoc postprocessing steps are used to compute the final
vector field and latent time based on aggregation and smoothing of the
independent velocity estimates for each gene which can dramatically
change the velocity estimates. In contrast, Pyro -Velocity is a fully
generative and probabilistic model that uses raw read counts. Our model
requires minimal preprocessing and postprocessing steps and outputs a
common shared time based on all the genes without additional aggregation
steps. It also provides uncertainty estimation based on posterior
samples. \textbf{b}, The main outputs of the Pyro -Velocity framework.
Visualization of single gene expression across inferred developmental
cell order using rainbow plots (right panel).}

\end{figure}%

\begin{figure}

\centering{

\includegraphics{figures/v2/Figure2.png}

}

\caption{\label{fig-model-results}Pyro -Velocity outperforms other
RNA-velocity methods and accurately predicts cell lineage trajectories.
\textbf{a,b}, RNA Velocity model comparisons from fully mature
peripheral blood mononuclear cells (PBMCs, t-SNE embedded plots, a) or
E15.5 mouse pancreas (UMAP embedded plots, \textbf{b-c}). From left to
right in panels \textbf{a} and \textbf{b}: 1. Cells colored by cell
type; 2. Stream plot of velocity vector field defined by scVelo; 3.
Stream plot of the posterior mean vector field from Pyro -Velocity; 4.
Pyro -Velocity visualization of cell dynamics for 3 arbitrarily selected
single cells; 5. Single-cell vector field uncertainty where each cell is
colored based on the scaled angular standard deviation to visualize
uncertainty based on n=30 iterated, posterior samples depicted as arrows
noted throughout our analysis; 6. Averaged vector field uncertainty
where each vector is colored by the averaged vector field uncertainty.
The scaled angular standard deviation range is {[}0 360{]}, where 0
corresponds to the highest confidence for a single direction, 360
corresponds to the highest uncertainty, and no directionality preference
based on 30 posterior samples (4-6). Vectors were projected into the
UMAP space using the scVelo API based on cosine similarity-derived
transition matrix. \textbf{c}, UMAP embedded cells colored by the
average cell shared time across 30 posterior samples. The Spearman's
correlation between Pyro -Velocity and Cytotrace is provided.
\textbf{d}, Pyro -Velocity automatic gene selection and ranking for
predicting important genes involved in pancreas development. \textbf{e},
Phase portraits of posterior samples average spliced and unspliced gene
expression (left), rainbow plots showing denoised spliced gene
expression (middle), and gene expression within UMAP embedded space.
Single dots correspond to individual cells (left, right) or vertical
lines (middle). Cell type annotation coloring is the same as in
\textbf{b}.}

\end{figure}%

\begin{figure}

\centering{

\includegraphics{figures/v2/Figure3.png}

}

\caption{\label{fig-lineage-tracing}Pyro -Velocity accurately predicts
cell fate choices and lineage trajectory in human LARRY barcoded
hematopoietic cells. Systematic benchmarking of cell fate prediction
methods - predicting uni-lineage cell fates (\textbf{a-b}, Model 1),
bifurcated lineages (\textbf{c}, Model 2), or all lineages together
(\textbf{d}, Model 2). Panels from left to right show on the same SPRING
embedding: 1. Cell types; 2. Clonal progression vector field derived
based on the observed clonal barcodes (\hyperref[sec-methods]{Methods});
3. scVelo recovered vector field; 4. Pyro -Velocity single cell level
vector field uncertainty (angular standard deviation from 30 posterior
samples); 5. Pyro -Velocity recovered vector field; vectors are colored
by the averaged vector field uncertainty; 6. Clonal fate potency score
using Cospar \citep{Wang2022-xb}; 7. scVelo recovered latent time; 8.
Pyro -Velocity inference of posterior mean shared time per cell. Cosine
similarity and Spearman's correlation for the different methods with the
clonal progression and fate potency are reported in the titles.}

\end{figure}%

\FloatBarrier
\clearpage

\newpage{}

\twocolumn

\section{Methods}\label{sec-methods}

\subsection{Model formulation}\label{sec-methods-model}

We assume the dynamical gene expression is determined by the RNA
splicing process, and infer the unspliced and spliced gene expression
level from the ordinary differential equation (ODEs) proposed in
velocyto \citep{La_Manno2018-lj} and scVelo \citep{Bergen2020-pj}
\begin{align}
\frac{du}{dt} &= \alpha(t) - \beta u, \quad u(0) = u_0 \label{eq-dudt}\\
   \frac{ds}{dt} &= \beta u - \gamma s, \quad s(0) = s_0, \label{eq-dsdt}
\end{align} where \(u(t), s(t)\) are the unspliced and spliced
expression levels of a gene at time \(t\) under a transcription rate
\(\alpha(t)\) with possible temporal dependence, splicing rate
\(\beta\), and degradation rate \(\gamma\). We specify this model to a
setting that depends on cell \(c\) and gene \(g\) as follows:
\begin{align}
\frac{du_{cg}}{dt} &= \alpha_{cg}(t) - \beta_{g} u_{cg}, \quad u_{cg}(0) = u_{cg}^{(0)} \label{eq-dudt}\\
   \frac{ds_{cg}}{dt} &= \beta_{g} u_{cg} - \gamma_{g} s_{cg}, \quad s_{cg}(0) = s_{cg}^{(0)} \label{eq-dsdt}.
\end{align} In the equation, the subscript \(c\) is the cell dimension,
\(g\) is the gene dimension, \(\left( u_{cg}(t), s_{cg}(t) \right)\) are
the unspliced and spliced expression functions given the change of time
per cell and gene. We restrict attention to piecewise-constant
\(\alpha_{cg}(t)\) to capture gene-specific activation and repression.
We take special care to model a gene- and cell-specific switching time
that marks a single transition from activation to repression by
introducing a Bernoulli variable \(k_{cg}\) to model unknown activation
state. We assume our cell-by-gene data-matrix arrive as observations of
Poisson-counts related to the solution of the above ODEs at unknown
times \(\tau_{cg}\), which is modeled as a relationship between an
unknown latent time shared across each cell, \(t_c\), and unknown
gene-specific time-offsets \(t_{0,g}\) where all read counts for a
single cell occurred at an unknown, but shared latent time \(t_c\).
These relative times are also used to parametrize the Bernoulli process
for \(k_{cg}\). Importantly, we recognize that the initial conditions
are in fact unknown.

We propose and study two models: Model 1 assumes that spliced and
unspliced concentrations are both 0 at time 0; Model 2 considers these
initial conditions as unknowns with a log-Normal prior distribution. In
general, the solution space of ODEs becomes much richer when considered
over a domain of initial conditions (as opposed to a single point);
indeed, this affords Model 2 much greater expressivity. For clarity, we
first present the generative framework for both models, then provide
further interpretation and intuition.

First, we introduce the generative model that describes the various
unobserved times: \begin{align}
  % unit lognormal t_c
  t_c &\sim \text{LogNormal}(0, 1) \\
  % gene-specific t_0
  t^{(0)}_{0,g} &\sim \text{LogNormal}(0, 1) \\
  % switching time
  \Delta \textrm{switching}_g &\sim \text{LogNormal}(0, 1) \\
  % gene-specific t_1
  t^{(1)}_{0,g} &= t^{(0)}_{0,g} + \Delta \textrm{switching}_g \\
  %cell-gene-specific activation state
  k_{cg} &\sim \text{Bernoulli}(\textrm{logits}=t_c - t^{(1)}_{0,g}) \\
  % cell-gene-specific latent time
  \tau_{cg} &= \text{softplus}(t_c - t^{(k_{cg})}_{0,g}).
\end{align} Here, \(\tau_{cg}\) represents the displacement of time per
cell and gene with \begin{align}
 \text{softplus}(t) :=  \log( 1 + e^t).
\end{align} Recall that \(t_c\) is the shared time per cell,
\(t^{(k_{cg})}_{0,g}\) is the gene-specific switching time. Each cell
and gene combination has its transcriptional state
\(k_{cg} \in \{ 0, 1 \}\), where \(0\) indicates the activation state
and \(1\) indicates the expression state. Each gene has two switching
times for representing activation and repression: \(t^{(0)}_{0,g}\) is
the first switching time corresponding to when the gene expression
starts to be activated, \(t^{(1)}_{0,g}\) is the second switching time
corresponding to when the gene expression starts to be repressed, and is
determined by the first switching time and the gene-specific switching
time \(\Delta \text{switching}_g\). The cell-gene-specific activation
state \(k_{cg}\) is a Bernoulli random variable with logits equal to the
difference between the cell's shared time \(t_c\) and the time
\(t^{(1)}_{0,g}\) when the gene expression starts to be repressed.

Next we introduce the priors for the splicing parameters (where the
activation rate \(\alpha\) depends on the activation state \(k_{cg}\)
from above): \begin{align}
  \alpha^{(0)}_g &\sim \text{LogNormal}(0, 1) \\
  \beta_g &\sim \text{LogNormal}(0, 1) \\
  \gamma_g &\sim \text{LogNormal}(0, 1) \\
  \alpha_{cg} &= \begin{cases}
    \alpha^{(0)}_g & \text{if } k_{cg} = 0 \\
    0 & \text{if } k_{cg} = 1
  \end{cases}
\end{align}
\textbf{Note that $\alpha^{(1)}$ is shared for all the genes, while ${\alpha^{(0)}}_g$ is learned independently for
each gene. MATT: this was in the old text, but I think $\alpha^{(1)}$ is no longer used based on conversations with Alvin?}

Now, we describe the priors for the initial conditions, noting that this
is the only difference between Model 1 and Model 2: \begin{align}
  \hat{u}^{(0)}_{cg}, \hat{s}^{(0)}_{cg} &\sim \begin{cases}
    (0, 0) & \text{Model 1} \\
    (\text{LogNormal}(0, 1), \text{LogNormal}(0, 1)) & \text{Model 2}
  \end{cases} \\
  u^{(0)}_{cg}, s^{(0)}_{cg} &= \begin{cases}
    \hat{u}^{(0)}_{cg}, \hat{s}^{(0)}_{cg} & \text{if } k_{cg} = 0 \\
    \textrm{ODESolve}\Big( \hat{u}^{(0)}_{cg}, \hat{s}^{(0)}_{cg}, \alpha^{(0)}_g, \beta_g, \gamma_g; \ T_0=0, T_1=\Delta \textrm{switching}_g \Big) & \text{if } k_{cg} = 1
  \end{cases}
\end{align}

We define the ODE solution at time \(\tau_{cg}\) as: \begin{equation}
    \hat{u}_{cg}, \hat{s}_{cg} = \text{ODESolve}\Big( u^{(0)}_{cg}, s^{(0)}_{cg}, \alpha_{cg}, \beta_g, \gamma_g; \ T_0=0, T_1=\tau_{cg} \Big).
\end{equation}

Next, we define the observation model that gives rise to the observed
counts as: \begin{align}
  \mu^{(u)}_c &= \sum_{g=1}^G {u}^{\text{(obs)}}_{cg}, \quad \mu^{(s)}_c = \sum_{g=1}^G {s}^{\text{(obs)}}_{cg} \\
  \sigma^{(u)}_c &= \sqrt{\frac{1}{G} \sum_{g=1}^G \left( u_{cg}^{\text{(obs)}} - \mu^{(u)}_c \right)^2} \\
  \sigma^{(s)}_c &= \sqrt{\frac{1}{G} \sum_{g=1}^G \left( s_{cg}^{\text{(obs)}} - \mu^{(s)}_c \right)^2} \\
  \eta^{(u)}_c &\sim \text{Normal}\Big(\mu^{(u)}_c, \ \sigma^{(u)}_c\Big) \\
  \eta^{(s)}_c &\sim \text{Normal}\Big(\mu^{(s)}_c, \ \sigma^{(s)}_c\Big) \\
  \hat{\mu}^{(u)}_c &= \sum_{g=1}^G \hat{u}_{cg}, \quad \hat{\mu}^{(s)}_c = \sum_{g=1}^G \hat{s}_{cg} \\
  \lambda^{(u)}_{cg} &= \log(\hat{u}_{cg}) + \log(\eta^{(u)}_{c}) - \log(\hat{\mu}^{(u)}_c) \\
  \lambda^{(s)}_{cg} &= \log(\hat{s}_{cg}) + \log(\eta^{(s)}_{c}) - \log(\hat{\mu}^{(s)}_c) \\
  \hat{u}^{\text{(obs)}}_{cg} &\sim \text{Poisson}\Big(\exp (\lambda^{(u)}_{cg})\Big) \\
  \hat{s}^{\text{(obs)}}_{cg} &\sim \text{Poisson}\Big(\exp (\lambda^{(s)}_{cg})\Big)
\end{align} Here, we use
\({u}^{\text{(obs)}}_{cg}, {s}^{\text{(obs)}}_{cg}\) to denote the
observed unspliced and spliced counts for cell \(c\) and gene \(g\). We
use \(\hat{u}^{\text{(obs)}}_{cg}, \hat{s}^{\text{(obs)}}_{cg}\) to
denote our generative model's prediction of these unspliced and spliced
expression levels. The generative process for modeling these observed
read counts given denoised gene transcript expression level
\(\hat{u}_{cg}, \hat{s}_{cg}\) considers the expected number of observed
reads for a given gene in a given cell as the number of transcripts
times the ratio of the cell's total reads to total transcripts.
\textbf{Improve descriptions of how noise is modeled in the observation model.}

\textbf{Need to update the analytic solutions, but first need to confirm the above is correct. Also, I recommend pushing all of the below analytic solutions to the appendix.}
The analytic solution of the differential equations to predict spliced
and unspliced gene expression given their parameters is derived by the
authors of scVelo and a theoretical RNA velocity study
\citep{Bergen2020-pj, Li2021-qa} and given in Eqs.
\ref{eq-solution-u}-\ref{eq-solution-s2}. \begin{align}
u\left(\tau^{\left(k_{c g}\right)}\right) 
  &= u_0^{\left(k_{c g}\right)}{ }_g e^{-\beta_g \tau^{\left(k_{c g}\right)}} 
  \nonumber \\
&\hskip -24pt + \frac{\alpha^{\left(k_{c g}\right)}}
  {\beta_g}\left(1-e^{-\beta_g \tau^{\left(k_{c g}\right)}}\right) 
  \label{eq-solution-u}\\
s\left(\tau^{\left(k_{c g}\right)}\right) 
  &= s_0^{\left(k_{c g}\right)} e^{-\gamma_g \tau^{\left(k_{c g}\right)}} 
  \nonumber \\
  &\hskip -24pt + \frac{\alpha^{\left(k_{c g}\right)}}{\gamma_g}
    \left(1-e^{-\gamma_g \tau^{\left(k_{c g}\right)}}\right) 
    \nonumber\\
  &\hskip -24pt + \frac{\alpha^{\left(k_{c g}\right)}-\beta_g u_0^{\left(k_{c g}\right)}}
    {\gamma_g-\beta_g}\left(e^{-\gamma_g \tau^{\left(k_{c g}\right)}}
    -e^{-\beta_g \tau^{\left(k_{c g}\right)}}\right), 
    \nonumber \\
  &\qquad \beta \neq \gamma \label{eq-solution-s} \\
s\left(\tau^{\left(k_{c g}\right)}\right) 
  &= {s_0^{\left(k_{c g}\right)}}_g e^{-\beta_g \tau^{\left(k_{c g}\right)}} \nonumber \\
  &\hskip -24pt +\frac{\alpha^{\left(k_{c g}\right)}}{\beta_g}
  \left(1-e^{-\beta_g \tau^{(k c g)}}\right) 
  \nonumber \\
  &\hskip -24pt -\left(\alpha^{\left(k_{c g}\right)}
  -\beta_g u_0^{\left(k_{c g}\right)}{ }_g\right) \tau^{\left(k_{c g}\right)} 
  e^{-\beta_g \tau^{\left(k_{c g}\right)}}, \nonumber \\
  &\qquad \beta = \gamma \label{eq-solution-s2}
\end{align}

To simplify these equations, consider the case when \(k_{cg} = 0\) and
\(\beta_g \neq \gamma_g\). Then, \begin{align}
u\left(\tau^{(0)}\right) &= {u_0^{(0)}}_g e^{-\beta_g \tau^{(0)}} \nonumber \\
  & \hskip -24pt + \frac{{\alpha^{(0)}}_g}{\beta_g}\left(1-e^{-\beta_g \tau^{(0)}}\right), 
  \label{eq-sol-usimp} \\
s\left(\tau^{(0)}\right) &= s_0^{(0)}{ }_g e^{-\gamma_g \tau^{(0)}} \nonumber \\
  & \hskip -24pt +\frac{{\alpha^{(0)}}_g}{\gamma_g}\left(1-e^{-\gamma_g \tau^{(0)}}\right) 
  \nonumber\\
  & \hskip -24pt +\frac{{\alpha^{(0)}}_g-\beta_g {u_0^{(0)}}_g}{\gamma_g-\beta_g}
  \left(e^{-\gamma_g \tau^{(0)}}-e^{-\beta_g \tau^{(0)}}\right). \label{eq-sol-ssimp}
\end{align} When \(k_{cg} = 0\) and \(\beta_g = \gamma_g\), then
\(u\left(\tau^{(0)}\right)\) has the same solution, and
\(s\left(\tau^{(0)}\right)\) becomes \begin{align}
s\left(\tau^{(0)}\right) &= s_0^{(0)}{ }_g e^{-\gamma_g \tau^{(0)}} \nonumber \\
  & \hskip -24pt +\frac{{\alpha^{(0)}}_g}{\gamma_g}
    \left(1-e^{-\gamma_g \tau^{(0)}}\right) \nonumber\\
  & \hskip -24pt - \left( {\alpha^{(0)}}_g-\beta_g {u_0^{(0)}}_g \right) 
    \tau^{(0)} e^{-\beta_g \tau^{(0)}}. \label{eq-sol-ssimp2}
\end{align} When \(k_{cg} = 1\) and \(\beta_g \neq \gamma_g\), then
\begin{align}
u\left(\tau^{(1)}\right) &=u_0^{(1)} g^{e^{-\beta_g \tau^{(1)}}}, \\
s\left(\tau^{(1)}\right) &=s_0^{(1)} e^{-\gamma_g \tau^{(1)}} \nonumber \\
  & \hskip -24pt +\frac{-\beta_g u_0^{(1)}}{\gamma_g-\beta_g}
  \left(e^{-\gamma_g \tau^{(1)}}-e^{-\beta_g \tau^{(1)}}\right).
\end{align} When \(k_{cg} = 1\) and \(\beta_g = \gamma_g\), then
\(u\left(\tau^{(1)}\right)\) has the same solution, and
\(s\left(\tau^{(1)}\right)\) becomes \begin{align}
s\left(\tau^{(1)}\right)=s_0^{(1)}{ }_g e^{-\gamma_g \tau^{(1)}}
  +\beta_g u_0^{(1)}{ }_g \tau^{(1)} e^{-\beta_g \tau^{(1)}}.
\end{align}

\subsection{Variational inference}\label{sec-methods-inference}

Given observations
\(\tilde{X}_{cg} = \left( u_{cg}^{obs}, s_{cg}^{obs} \right)\), we would
like to compute the posterior distribution over the random variables
\begin{align}
\theta &= \left( t_{c}, \eta_{c}^{(u)}, \eta_{c}^{(s)} \right), \\
\phi &= \left( {t_{0}^{(0)}}_g, \Delta \text{switching}_{g}, 
          {\alpha^{(0)}}_{g}, \beta_{g}, \gamma_{g} \right),
\end{align} but exact Bayesian inference is intractable in this model.
We use Pyro to automatically integrate out the local discrete latent
variables \(k\), which is defined as the cell and gene transcriptional
state (see above), and approximate the posterior over the remaining
latent variables using variational inference
\citep{Bingham2018-id, Kucukelbir2016-bk}, which converts intractable
integrals into optimization problems that can be solved with
off-the-shelf tools. In variational inference, we maximize a tractable
bound (the evidence lower bound (ELBO)) on the Kullback-Leibler (KL)
divergence \[
\text{KL}\left( q_{\phi}(\theta, \psi) \mathrel{\Vert} 
  p(\theta, \psi \vert \tilde{X}_{cg}) \right)
\] between a parametric family of tractable probability distributions q
and the intractable true posterior \begin{align}
\varphi^* &= \operatorname{argmax}_{\varphi} \text{ELBO} \nonumber \\
&= \operatorname{argmax}_{\varphi} \bigg\{ E_{q_{\varphi}} 
     \bigg[ \log \left(p\left(\tilde{X}_{c g}, \theta, \psi \right) \right) \nonumber\\
&\qquad\qquad\qquad\qquad -\log \left(q_{\varphi}(\theta, \psi)\right) \bigg] \bigg\}
\end{align} We approximate our model's posterior distribution
\(p( \theta, \psi \vert
\tilde{X})\) with a tractable family of probability distributions
\(q_{\psi}(\theta, \phi) = q_{\phi_1}(\theta) q_{\phi_2}(\psi)\), where
\(q_{\phi_1}(\theta)\) is a product of independent univariate Gaussian
distributions with learnable location and scale parameters and
\(q_{\phi_2}(\psi)\) is a multivariate Gaussian distribution with a
learnable location parameter and a learnable low-rank covariance matrix.
We solve the resulting stochastic optimization problem using a version
of Automatic Differentiation Variational Inference (ADVI)
\citep{Kucukelbir2016-bk}: we obtain gradient estimates by
differentiating a Monte Carlo estimate of the ELBO and update the
variational parameters using stochastic gradient ascent. Our
implementation and experiments use generic variational families and
Monte Carlo ELBO and gradient estimators provided by Pyro
\citep{Bingham2018-id}, an open source library for probabilistic machine
learning, and a version built into Pyro of the adaptive stochastic
gradient ascent algorithm Adam augmented with gradient clipping to
enhance numerical stability during training.

\subsection{Model training}\label{sec-methods-training}

For the pancreas, PBMC, and uni-fate and bi-fate LARRY single-cell data,
since the data dimension is relatively small, we run Model 1 and Model 2
with a minimum of 100 epochs and a maximum of 4000 epochs. For each
epoch, we input spliced and unspliced read counts from all the cells and
determine the convergence condition with an early stopping strategy in
which the patience is set to 45 and consumes one patience if the minimal
ELBO improvement of training data per epoch is lower than \(10^{-4}\) of
previous loss. The learning rate is set to be \(10^{-2}\) with a decay
rate of \(0.11/4000\) per epoch. For the multi-fate LARRY dataset, since
this is a large dataset with over \(4 \times
10^4\) cells, we use mini batches of cells to train Model 1 and Model 2
with a minimum of 100 epochs and a maximum of 1000 epochs. Specifically,
the batch size was set to 4000 cells for both models with an early
stopping patience 45 and consume one patience if the minimal ELBO
improvement of training data per epoch lower than \(10^{-3}\) of
previous loss. The learning rate is set to be \(10^{-2}\) with a decay
rate of \(0.11/1000\) per epoch. All models were trained on a machine
with an NVIDIA A100 GPU and the CentOS 7 operating system.

\subsection{Posterior
prediction}\label{sec-methods-posterior-prediction}

To benchmark Pyro -Velocity performance in predicting cell fate, we
generated the posterior samples measurement
\(x_{cg}=\left(u\left(\tau^{(k_{cg})}\right),
s\left(\tau^{(k_{cg})}\right)\right)\) or
\(x_{cg}=\left(u_{cg}, s_{cg} \right)\), \(t_c\), \(\beta_g\),
\(\gamma_g\) from the same single cell RNA-seq using \(N=30\) Monte
Carlo samples from the posterior predictive distribution following
\(p(x_{cg} \vert \theta, \psi) p(\theta, \psi) \approx p(x_{cg} \vert \theta,
\psi) q_{\phi}(\theta, \psi)\). \(x_{cg}\) can be posterior samples of
either denoised gene expression (used for phase portraits and vector
field-based trajectory inference) or raw read counts (used for gene
selection). Then we calculated the posterior samples of RNA velocity as
\(v_{cg}=\frac{ds(\tau^{(k_{cg})})}{d \tau^{(k{cg})}}\) based on
posterior samples measurement of denoised gene expression \(x_{cg}\) and
\(\beta_{g}\), \(\gamma_{g}\).

\subsection{Prioritization of cell fate
markers}\label{sec-methods-fate-markers}

We prioritize the cell fate markers using two metrics: First, the
Pearson correlation between each gene's posterior mean of the denoised
spliced expression and posterior mean of the shared time; Second, the
negative mean absolute errors of each gene's observed spliced, and
unspliced read counts with posterior predictive samples of spliced and
unspliced raw read counts, i.e.
\(\frac{1}{N n_c} \sum_i \sum_c (x_{icg}-\tilde{X}_{cg})\), where \(N\)
is the posterior sample number that is set to \(30\), \(i\) is the
posterior sample index, and \(n_c\) is the cell number. We first select
the top \(300\) genes with the highest negative mean absolute errors and
then rank the \(300\) genes based on the most positively correlated
genes and the least negatively correlated genes. We use the same
strategy for scVelo to rank the markers by the model likelihood to get
the top 300 genes and then prioritize these genes by Pearson's
correlation between scVelo latent time and normalized expression.

\subsection{Single-cell data
preprocessing}\label{sec-methods-preprocessing}

We used scanpy and scVelo to handle the data input and output; thus,
both h5ad and loom files generated by velocyto and kallisto
\citep{Melsted2021-ap} are supported. The fully mature PBMC dataset was
processed with the same procedure proposed in a review paper
\citep{Bergen2021-qz}
(\url{https://scvelo.readthedocs.io/perspectives/Perspectives/}). We
reproduced this procedure using the scVelo package and raw read counts
of the same top three dynamical genes NKG7, IGHM, and GNLY with the best
likelihoods. The pancreas dataset was processed with scVelo using the
following options

\begin{Shaded}
\begin{Highlighting}[]
\ImportTok{import}\NormalTok{ scvelo}

\NormalTok{scvelo.pp.filter\_and\_normalize(}
\NormalTok{  adata}\OperatorTok{=}\NormalTok{adata,}
\NormalTok{  min\_shared\_counts}\OperatorTok{=}\DecValTok{30}\NormalTok{, }
\NormalTok{  n\_top\_genes}\OperatorTok{=}\DecValTok{2000}\NormalTok{,}
\NormalTok{)}
\NormalTok{scvelo.pp.moments(}
\NormalTok{  adata,}
\NormalTok{  n\_pcs}\OperatorTok{=}\DecValTok{30}\NormalTok{,}
\NormalTok{  n\_neighbors}\OperatorTok{=}\DecValTok{30}\NormalTok{,}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

The same top variable genes with raw spliced and unspliced read counts
were used as input for the Pyro -Velocity model. The original LARRY
dataset of in vitro Hematopoiesis containing \(130,887\) cells was first
filtered to remove cells without LARRY barcoding. \(49,302\) cells were
recovered after this step with at least one LARRY barcode. For
simplicity, we termed this filtered dataset with multiple cell fate
(multi-fate) as the full dataset. Based on this dataset, we created two
datasets with uni-fate progression toward monocyte or neutrophil based
on the lineage LARRY barcodes and time information. Namely, we selected
sets of cells with a single LARRY barcode, spanning three time points
(day 2, 4, 6), and all the cells from the last time point (day 6) belong
to a unique cell type (either monocyte or neutrophil). The two uni-fate
datasets were combined to represent the bi-fate LARRY dataset. The
multi-fate full dataset was processed using the same options as the
pancreas dataset; the rest of the uni-fate and bi-fate datasets were
processed using the following parameters

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{scvelo.pp.filter\_and\_normalize(}
\NormalTok{  adata}\OperatorTok{=}\NormalTok{adata,}
\NormalTok{  n\_top\_genes}\OperatorTok{=}\DecValTok{2000}\NormalTok{,}
\NormalTok{  min\_shared\_counts}\OperatorTok{=}\DecValTok{20}\NormalTok{,}
\NormalTok{)}
\NormalTok{scvelo.pp.moments(adata)}
\end{Highlighting}
\end{Shaded}

\subsection{scVelo model}\label{sec-methods-scvelo}

We benchmarked the dynamical RNA velocity model implemented in scVelo
\texttt{v0.2.4} for the pancreas and the four LARRY datasets using the
same user options

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{scvelo.tl.recover\_dynamics(}
\NormalTok{  data}\OperatorTok{=}\NormalTok{adata, }
\NormalTok{  n\_jobs}\OperatorTok{=}\DecValTok{30}\NormalTok{,}
\NormalTok{)}
\NormalTok{scvelo.tl.velocity(}
\NormalTok{  data}\OperatorTok{=}\NormalTok{adata, }
\NormalTok{  mode}\OperatorTok{=}\StringTok{"dynamical"}\NormalTok{,}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

Then, we tested a set of user options, including the neighboring cell
numbers and the top variable gene numbers, in the pancreas dataset to
explore the stability of the scVelo dynamical model. For the fully
mature PBMC dataset, we followed the notebook proposed by the original
authors \url{https://scvelo.readthedocs.io/perspectives/Perspectives/},
i.e., we used the stochastic RNA velocity model implemented in scVelo
with the top three likelihood genes. The latent time from scVelo was
computed using their provided function

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{scvelo.tl.latent\_time(}
\NormalTok{  data}\OperatorTok{=}\NormalTok{adata,}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\subsection{Trajectory
inference}\label{sec-methods-trajectory-inference}

\subsubsection{Velocity vector field}\label{velocity-vector-field}

The velocity-based vector fields were generated using the scVelo
function call

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{scvelo.tl.velocity\_graph(adata)}
\NormalTok{scvelo.tl.velocity\_embedding(}
\NormalTok{  adata, }
\NormalTok{  vkey}\OperatorTok{=}\StringTok{\textquotesingle{}velocity\textquotesingle{}}\NormalTok{,}
\NormalTok{  basis}\OperatorTok{=}\StringTok{\textquotesingle{}embed\textquotesingle{}}\NormalTok{,}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

We used the default options for projecting the vector fields from scVelo
models. Unlike scVelo, Pyro -Velocity uses statistics derived from
posterior samples of the denoised spliced gene expression and posterior
samples of the velocity estimation for building the cell state
transition matrix estimates using cosine similarity. Pyro -Velocity uses
the same projection method as scVelo for projecting the transition
matrix into the two-dimensional vector field on the user-provided
embedding space.

\subsubsection{Clonal progression vector
field}\label{clonal-progression-vector-field}

No method exists for projecting clonal cell fates from the LARRY data
and associated subsets thereof into a vector field. Thus, we designed a
strategy for projecting clonal progression vector fields. Given cells
sharing the same LARRY barcode, we considered the vectors connecting the
embedding centroid of cells at consecutive time points and then
leveraged Gaussian smoothing of nearest centroid directions to represent
the clonal progression vector field informed by LARRY barcodes on a
regular grid over the cell manifold.

\subsection{Cell fate uncertainty
estimation}\label{sec-methods-fate-uncertainty}

Pyro -Velocity evaluates cell fate uncertainty using four statistics.
The first is the uncertainty of shared time, which is evaluated using
the standard deviation of posterior samples of \(t_c\) for each cell.
The second statistic is the angular standard deviation
\citep{Berens2009-fu} across posterior samples of the velocity-based
vector field. The value of this statistic was mapped linearly to the
range \([0, 360]\) for interpretability. The third statistic is the
Rayleigh test \citep{The-Astropy-Collaboration2022-aw} for posterior
samples of the velocity-based vector fields for each cell, whose
p-values have been corrected for multiple tests using Benjamini-Hochberg
FDR methods; if FDR is less than \(0.05\), the cell is considered to
have one statistically significant future direction. The second and
third circular statistics were implemented using the Astropy package
(\texttt{v5.1}) \citep{The-Astropy-Collaboration2022-aw}. The last
statistic is state uncertainty which evaluates the mean Euclidean
distance between posterior mean and posterior samples of the raw read
count prediction for each cell \begin{align*}
\frac{1}{N} \sum_{i} \sqrt{\sum_{g} \left(x_{icg} - \frac{1}{N}\sum_i x_{icg}\right)^2}.
\end{align*}

\subsection{Cospar}\label{sec-methods-cospar}

We computed the
\href{https://github.com/AllonKleinLab/cospar/releases/tag/v0.1.9}{Cospar
\texttt{v0.1.9}} \citep{Wang2022-xb} fate potency score for the above
processed multi-fate LARRY dataset, and selected the subset of the fate
potency scores for evaluating bi-fate and uni-fate datasets. We used the
following function calls to achieve this

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{cospar.tmap.infer\_Tmap\_from\_multitime\_clones(}
\NormalTok{  adata\_orig}\OperatorTok{=}\NormalTok{adata, }
\NormalTok{  compute\_new}\OperatorTok{=}\VariableTok{True}\NormalTok{,}
\NormalTok{  clonal\_time\_points}\OperatorTok{=}\NormalTok{[}\DecValTok{2}\NormalTok{,}\DecValTok{4}\NormalTok{,}\DecValTok{6}\NormalTok{], }
\NormalTok{  later\_time\_point}\OperatorTok{=}\DecValTok{6}\NormalTok{,}
\NormalTok{  smooth\_array}\OperatorTok{=}\NormalTok{[}\DecValTok{20}\NormalTok{,}\DecValTok{15}\NormalTok{,}\DecValTok{10}\NormalTok{],}
\NormalTok{  sparsity\_threshold}\OperatorTok{=}\FloatTok{0.2}\NormalTok{, }
\NormalTok{  max\_iter\_N}\OperatorTok{=}\DecValTok{3}
\NormalTok{)}
\NormalTok{cospar.pl.fate\_potency(}
\NormalTok{  adata}\OperatorTok{=}\NormalTok{adata, }
\NormalTok{  used\_Tmap}\OperatorTok{=}\StringTok{\textquotesingle{}transition\_map\textquotesingle{}}\NormalTok{,}
\NormalTok{  map\_backward}\OperatorTok{=}\VariableTok{True}\NormalTok{,}
\NormalTok{  method}\OperatorTok{=}\StringTok{\textquotesingle{}norm{-}sum\textquotesingle{}}\NormalTok{,}
\NormalTok{  color\_bar}\OperatorTok{=}\VariableTok{True}\NormalTok{,}
\NormalTok{  fate\_count}\OperatorTok{=}\VariableTok{True}
\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\subsection{Cytotrace model}\label{sec-methods-cytotrace}

We implemented Cytotrace in Python based on the R version
\citep{Gulati2020-xq} for benchmarking Pyro -Velocity and scVelo's
shared (latent) time prediction in the Python environment. This
implementation is available in the
\href{https://github.com/pinellolab/pyrovelocity/blob/main/src/pyrovelocity/analysis/cytotrace.py}{Pyro -Velocity
GitHub repository}.

\subsection{Model evaluation and
selection}\label{sec-methods-model-selection}

Pyro -Velocity provides two model formulations as described above. Users
can compare their predictive performance to decide which of the two
models to use for a given dataset. Namely, the mean absolute errors of
posterior samples prediction of spliced and unspliced raw read counts
\(x_{icg}\) from its observation \(\tilde{X}_{cg}\) per gene and cell
combinations \begin{align*}
\frac{1}{N_i N_c N_g} \sum_{i=1}^{N_i} \sum_{c=1}^{N_c} \sum_{g=1}^{N_g} 
  \vert x_{icg} - \tilde{X}_{cg} \vert,
\end{align*} where \(i\) is the posterior sample index, \(N_i\) is the
posterior sample number, \(N_c\) is the cell number, \(N_g\) is the gene
number. Then it is possible to evaluate the cell fate inference
performance of the models based on two metrics:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  the consistency between the velocity vector field and the \emph{clonal
  progression vector field} using grid-based cosine similarity; 2.
  Spearman's correlation of velocity-based shared (latent) time with
  Cospar fate potency and the silver standard Cytotrace score. Both
  metrics are implemented in scipy.
\end{enumerate}

\section{Data availability}\label{data-availability}

All the data presented in this manuscript were collected from previously
published studies. The fully mature PBMC, pancreas datasets are
downloaded and loaded through the scVelo using
\texttt{scvelo.datasets.pancreas()}, and
\texttt{scvelo.datasets.pbmc68k()}. The LARRY inDrop sequencing data was
processed by the LARRY authors using dropEst \citep{Petukhov2018-nr} and
velocyto. This dataset can be accessed at
\url{https://figshare.com/articles/dataset/larry_invitro_adata_sub_raw_h5ad/20780344}.

\section{Code availability}\label{code-availability}

We developed our Pyro -Velocity models with Pyro (version 1.6.0)
probabilistic programming language and PyTorch (version 1.8.1).
Pyro -Velocity source code is available at
\url{https://github.com/pinellolab/pyrovelocity}. The version of
Pyro -Velocity (v0.1.0) used to generate the results presented in this
manuscript has been deposited on Zenodo:
\url{https://zenodo.org/record/7072876\#.Yx-2e-zMLxg}.

\section*{Acknowledgments}\label{acknowledgments}
\addcontentsline{toc}{section}{Acknowledgments}

This project has been made possible in part by grant number 2022-249212
from the Chan Zuckerberg Initiative DAF, an advised fund of Silicon
Valley Community Foundation. L.P. received support from the U.S. NIH
(R35 HG010717) and Silicon Valley Community Foundation (2022-249212).
D.M.L is supported by R01CA226926, R01 CA215118, R01CA154923, R24
OD031955, CureSearch Acceleration Initiative, and the MGH Scholars
Award. We thank the Pinello and Langenau labs members for helpful
discussions.

\section*{Author contributions}\label{author-contributions}
\addcontentsline{toc}{section}{Author contributions}

L.P., Q.Q., and E.B. conceived the work. Q.Q. and E.B created the
software. Q.Q. analyzed the data, and prepared the figures and tables.
L.P and D.L. supervised the project. L.P., D.L., E.B., G.L.M helped in
designing experiments and analyses. All the authors contributed to the
writing of the manuscript.

\section*{Competing financial interests
statement}\label{competing-financial-interests-statement}
\addcontentsline{toc}{section}{Competing financial interests statement}

L.P. has financial interests in Edilytics, Inc., Excelsior Genomics, and
SeQure Dx, Inc.~L.P.'s interests were reviewed and are managed by
Massachusetts General Hospital and Partners HealthCare in accordance
with their conflict of interest policies.

\FloatBarrier
\clearpage

\newpage{}

\onecolumn

\section{Extended Data Figures}\label{extended-data-figures}

\begin{suppfig}

\centering{

\includegraphics{figures/v2/Supplementary-Figure1.pdf}

}

\caption{\label{extfig-graphical-models}Graphical representations detail
the generative process of the Pyro -Velocity models and highlight their
differences. \textbf{a}, Pyro -Velocity Model 1. Each rectangle
represents one data dimension (e.g., gene or cell). The arrows represent
the directed hierarchical dependency among random variables. The white
circles correspond to the modeled random variables. The probability
distributions of these variables are denoted on the right legend. The
gray circles correspond to intermediate results computed from random
variables. In this model, the kinetics random variables for the cell
shared time and gene are modeled with two constraints: 1. the shared
time is strictly larger than the gene-specific activation time
\(t_{0}^{(0)}\), 2. the gene expression in the activate state starts
with the initial condition \((u_{0}=0,\,s_{0}=0)\), as in previous
methods\citep{La_Manno2018-lj, Bergen2020-pj}
(\hyperref[sec-methods]{Methods}). \textbf{b}, Pyro -Velocity Model 2.
This figure follows the same graphical convention as in \textbf{a}, In
this model \(t_{0}^{(0)}\) is independent from the shared time, thus
allowing a time lag for the gene activation phase. In addition, a
positive starting basal gene expression level \((u_{0},\,s_{0})\)
different than \((u_{0}=0,\,s_{0}=0)\) is allowed.
(\hyperref[sec-methods]{Methods}). \textbf{c}, Schematic illustrating
how the constraints proposed in Model 1 (left) and Model 2 (right) are
reflected in the modeling of individual gene expression. The x-axis
corresponds to the shared time, the y-axis corresponds to the posterior
mean of the spliced expression level, the vertical lines with arrows
show the start and end time of the observed cell sampling, t0 represents
the gene-specific activation time \(t_{0}^{(0)}\), the red lines
highlight the differences for the potential gene-specific time lag when
\(t_{0}^{(0)} > {\min}\!⁡(t)\), and the initial spliced gene expression.
The same differences can also be illustrated for the unspliced gene
expression profiles (not shown).}

\end{suppfig}%

\begin{suppfig}

\centering{

\includegraphics{figures/v2/Supplementary-Figure2.png}

}

\caption{\label{extfig-model-results}Extended version of
Fig.~\ref{fig-model-results} for evaluating Pyro -Velocity (Model 1) and
scVelo on the PBMC and pancreas datasets. \textbf{a}, Left figure shows
the scVelo vector field using the default user options, right figure
shows the evaluation of the scVelo vector field using different
combinations of user options within Epsilon cells. \textbf{b},
Pyro -Velocity additional single cell level uncertainty diagnostic plots
on the fully mature PBMC dataset, from left to right: 1. standard
deviation of posterior samples of shared time, 2. Rayleigh test of
posterior samples of the vector field (the title reports false discovery
rate after p-value correction for multiple tests across cells), 3.
transcriptional state uncertainty of predicted raw read count per cell.
The contour lines show the top 10\% of the most uncertain cells.
\textbf{c}, For the pancreas dataset, the figures show from left to
right: UMAP rendering of 1. Cytotrace, 2. scVelo latent time, 3.
Posterior mean of Pyro -Velocity shared time averaged across 30
posterior samples, 4. Top 50 markers overlapping between Pyro -Velocity
and scVelo (\hyperref[sec-methods]{Methods}, Supplementary Table 1-2),
5. The same uncertainty diagnostic plots per cell as in \textbf{b}.
\textbf{d}, The same plot as Fig.~\ref{fig-model-results}d, e. Selection
of the top negatively correlated genes with the cell shared time based
on the posterior mean on denoised spliced expression and negative mean
absolute error of gene expression prediction
(\hyperref[sec-methods]{Methods}). \textbf{f}, Phase portraits, rainbow
plots, and UMAP rendering of posterior mean of gene expression for the
genes selected in \textbf{e}.}

\end{suppfig}%

\begin{suppfig}

\centering{

\includegraphics{figures/v2/Supplementary-Figure3.png}

}

\caption{\label{extfig-model-2-result}Extension of
Fig.~\ref{fig-model-results}, Pyro -Velocity (Model 2) outperforms
scVelo on PBMC and pancreas datasets. Same format as in
Fig.~\ref{fig-model-results} and
Extended Data Fig.~\ref{extfig-model-results}.}

\end{suppfig}%

\begin{suppfig}

\centering{

\includegraphics{figures/v2/Supplementary-Figure4.png}

}

\caption{\label{extfig-lineage-tracing}Extension of
Fig.~\ref{fig-lineage-tracing}, Pyro -Velocity (Model 1 for uni-fate
data, Model 2 for bi-fate and multi-fate data) outperforms Cytotrace on
scRNA-seq datasets with lineage barcoding information (LARRY dataset).
\textbf{a}, SPRING embedding rendering of Cytotrace scores, the title
reports the Spearman's correlation between the Cytotrace and Cospar fate
potency scores, \textbf{b}, Pyro -Velocity additional single cell level
uncertainty diagnostic plots based on the multi-fate LARRY dataset, from
left to right: 1. Standard deviation of posterior samples of shared time
per cell, 2. Rayleigh test of posterior samples of vector field per cell
(the title reports the false discovery rate after p-value correction for
multiple tests across cells), 3. Transcriptional state uncertainty of
predicted raw read count per cell. The contour lines show the top 10\%
of the most uncertain cells (\hyperref[sec-methods]{Methods}),
\textbf{c}, Pairwise Spearman's correlation between Cytotrace/Cospar
fate potency scores and velocity methods for the four different
benchmarking datasets selected based on the lineage barcoding
information from LARRY data, \textbf{d}, SPRING embedding rendering of
posterior mean of shared time across 30 posterior samples, the title
shows the Spearman's correlation of mean shared time with Cytotrace.
\textbf{e}, Selection of the top positively correlated genes with the
cell shared time based on the posterior mean of spliced expression.
\textbf{f}, Phase portraits, rainbow plots, and UMAP rendering with
splicing gene expression level for the genes selected in \textbf{e}.}

\end{suppfig}%

\begin{suppfig}

\centering{

\includegraphics{figures/v2/Supplementary-Figure5.png}

}

\caption{\label{extfig-lineage-tracing-2}Extension of
Fig.~\ref{fig-lineage-tracing} presenting an extended comparison of
Pyro -Velocity and scVelo on scRNA-seq datasets with lineage barcoding
information (LARRY dataset). \textbf{a}, uni-fate Monocyte lineage
analysis with Pyro -Velocity Model 2, \textbf{b}, uni-fate Neutrophil
lineage analysis with Pyro -Velocity Model 2, \textbf{c}, Neutrophil and
Monocyte bi-fate lineage analysis with Pyro -Velocity Model 1, and
\textbf{d}, multi-fate lineage analysis with Pyro -Velocity Model 1. The
panel descriptions are the same as in Fig.~\ref{fig-lineage-tracing}.}

\end{suppfig}%





\end{document}
